fastparquet/core.py

```diff
@@ -98,7 +98,16 @@ def read_data_page(f, helper, header, metadata):
     io_obj = encoding.Numpy8(np.frombuffer(memoryview(raw_bytes),
                                            dtype=np.uint8))
 
-    definition_levels, num_nulls = read_def(io_obj, daph, helper, metadata)
+    if skip_nulls and not helper.is_required(metadata.path_in_schema[-1]):
+        num_nulls = 0
+        definition_levels = None
+        io_obj.loc += 6
+        n = daph.num_values // 64
+        while n:
+            io_obj.loc += 1
+            n //= 128
```

Comment:
This logic seems somewhat complex.  Is this something that should be separated so that it can be tested independently?  (genuine question, not a nit)